lib/util/source-code.js

```diff
@@ -223,14 +224,110 @@ class SourceCode extends TokenStore {
     }
 
     /**
+     * Gets all comments before the given node or token.
+     * @param {ASTNode|token} nodeOrToken The AST node or token to get the comments for.
+     * @returns {Object} The list of comments indexed by their position.
+     * @public
+     */
+    getCommentsBefore(nodeOrToken) {
+        if (this._commentCache.has("getCommentsBefore", nodeOrToken)) {
+            return this._commentCache.get("getCommentsBefore", nodeOrToken);
+        }
+
+        const comments = [];
+
+        /**
+         * The Program node's range[0] corresponds to the start value of the
+         * first node (ignoring comment tokens at the beginning of the program).
+         * As a result, calling this.getTokenBefore() with the Program node
+         * will return comments at the beginning of a file.
+         */
+        if (nodeOrToken.type === "Program") {
+            return comments;
+        }
+
+        let currentToken = this.getTokenBefore(nodeOrToken, { includeComments: true });
+
+        while (currentToken && astUtils.isCommentToken(currentToken)) {
+            comments.unshift(currentToken);
+            currentToken = this.getTokenBefore(currentToken, { includeComments: true });
+        }
+
+        this._commentCache.set("getCommentsBefore", nodeOrToken, comments);
+        return comments;
+    }
+
+    /**
+     * Gets all comments after the given node or token.
+     * @param {ASTNode|token} nodeOrToken The AST node or token to get the comments for.
+     * @returns {Object} The list of comments indexed by their position.
+     * @public
+     */
+    getCommentsAfter(nodeOrToken) {
+        if (this._commentCache.has("getCommentsAfter", nodeOrToken)) {
+            return this._commentCache.get("getCommentsAfter", nodeOrToken);
+        }
+
+        const comments = [];
+
+        /**
+         * The Program node's range[1] corresponds to the end value of the
+         * last node (ignoring comment tokens at the end of the program).
+         * As a result, calling this.getTokenAfter() with the Program node
+         * will return comments at the end of a file.
+         */
+        if (nodeOrToken.type === "Program") {
+            return comments;
```

Comment:
Hmm, that seems like a bug to me. I think `start` and `range[0]` are supposed to always be the same.